### PR TITLE
Adding SuperfluidLoaderLib to provide host address for any network

### LINKED ISSUE #1367: Create SuperfluidLoaderLib providing the host address on any supported network

### PR DESCRIPTION
**Pull request by shreyaspapi**
- Adds the SuperfluidLoaderLib Solidity library

fixes #1367

### CODE REVIEW

**Comment by shreyaspapi** [comment]
In addition, we can add an enum called `ChainId` to `SuperfluidLoaderLib` for improved network ID representation:

```
enum ChainId {
    MAINNET,
    GOERLI_TESTNET,
    BSC_MAINNET,
    ARBITRUM_TESTNET,
    OPTIMISM_TESTNET,
    UNKNOWN
}
```

This change will increase the size of the code but can significantly improve readability.

**Comment by 0xdavinchee** [comment]
> In addition, we can add an enum called `ChainId` to `SuperfluidLoaderLib` for improved network ID representation:
> 
> ```
> enum ChainId {
>     MAINNET,
>     GOERLI_TESTNET,
>     BSC_MAINNET,
>     ARBITRUM_TESTNET,
>     OPTIMISM_TESTNET,
>     UNKNOWN
> }
> ```
> 
> This change will increase the size of the code but can significantly improve readability.

gm! Thank you for your contribution.

Some notes:

Using enums like this will mean MAINNET = 0, GOERLI_TESTNET = 1, etc. so this would not work. However, I do think adding these as constant fields in the library would significantly improve readability and worth the slight overhead.

e.g.

```
uint256 private constant MAINNET = 1;
uint256 private constant GOERLI_TESTNET = 5;
```
etc.

**Comment by shreyaspapi** [comment]
> Using enums like this will mean MAINNET = 0, GOERLI_TESTNET = 1, etc. so this would not work. However, I do think adding these as constant fields in the library would significantly improve readability and worth the slight overhead.

I was thinking of having - `getChainId` and `chainIdToHost` functions.

> ```
> uint256 private constant MAINNET = 1;
> uint256 private constant GOERLI_TESTNET = 5;
> ```

However, this approach appears to be better and simpler. I will proceed with making this change.

**Comment by 0xdavinchee** [comment]
gm!
Taking another look today and a few changes, these are purely to maintain consistency:
1. Please move `SuperfluidLoaderLib.sol` to the `libs` folder under `contracts`
2. Please add the host address for `BASE_GOERLI`: `0x507c3a7C6Ccc253884A2e3a3ee2A211cC7E796a6` and `ETH_SEPOLIA`: `0x109412E3C84f0539b43d39dB691B08c90f58dC7c`
3. Lastly, please change the variable names of the addresses to match the `uppercaseName` in the metadata file [here](https://github.com/superfluid-finance/metadata/blob/master/networks.json).

**Comment by shreyaspapi** [comment]
> gm! Taking another look today and a few changes, these are purely to maintain consistency:

@0xdavinchee GM!

> 1. Please move `SuperfluidLoaderLib.sol` to the `libs` folder under `contracts`
> 2. Please add the host address for `BASE_GOERLI`: `0x507c3a7C6Ccc253884A2e3a3ee2A211cC7E796a6` and `ETH_SEPOLIA`: `0x109412E3C84f0539b43d39dB691B08c90f58dC7c`
> 3. Lastly, please change the variable names of the addresses to match the `uppercaseName` in the metadata file [here](https://github.com/superfluid-finance/metadata/blob/master/networks.json).

Done ✅

**Review by 0xdavinchee — APPROVED** [review]
LGTM 🎉

**Review by hellwolf — APPROVED** [review]
@0xdavinchee could you piggy back a fix somewhere for this where SuperfluidLoaderLibrary is solcover ignored?

**Comment by hellwolf** [comment]
@0xdavinchee oh too bad, codecov/patch  is mandatory :) but maybe we can modify the branch for shreyaspapi?

**Comment by hellwolf** [comment]
> @0xdavinchee oh too bad, codecov/patch is mandatory :) but maybe we can modify the branch for shreyaspapi?

I think another standard workflow is to suggest the PR author for code changes.

**Comment by shreyaspapi** [comment]
@hellwolf, should I add `SuperfluidLoaderLib` to `solcover.js` skipFiles configuration to make the CI skip it?

Or should we make small uint test?

**Comment by hellwolf** [comment]
> @hellwolf, should I add `SuperfluidLoaderLib` to `solcover.js` skipFiles configuration to make the CI skip it?
> 
> Or should we make small uint test?

This library doesn't seem to be worth testing locally. What we could do is move this to apps/ folder, where deployment-specifics seem more appropriate. Additionally it could then be added to solcover ignore list, where these app/* is ignored already:

```
        // we skip the coverage for the SuperAppBase contracts because
        // we override the functions in child contracts
        "apps/SuperAppBase.sol",
        "apps/SuperAppBaseCFA.sol",
```

**Review by hellwolf — APPROVED** [review]
thank you!

File: packages/ethereum-contracts/.solcover.js [hunk 1]
```diff
@@ -10,6 +10,7 @@ module.exports = {
         // we override the functions in child contracts
         "apps/SuperAppBase.sol",
         "apps/SuperAppBaseCFA.sol",
+        "apps/SuperfluidLoaderLibrary.sol",
```
Comment (shreyaspapi) [inline packages/ethereum-contracts/.solcover.js:13]: It's still not ignoring the file. Does the path need to be `contracts/apps/SuperfluidLoaderLibrary.sol`?

Comment (0xdavinchee) [reply packages/ethereum-contracts/.solcover.js:13]: hmm, that is weird, the path is fine as is, I am guessing this is coming from the foundry coverage.
we have a coverage report generated by hardhat which uses the `.solcover.js` file and we also have one generated by foundry which we are currently cleaning up ad-hoc.

To fix it, you need to modify `tasks/coverage-cleanup.sh`, you can add: 'packages/ethereum-contracts/contracts/apps/SuperfluidLoaderLibrary.sol' to the line where we remove stuff:
`lcov -r lcov.info 'packages/ethereum-contracts/contracts/mocks/*' 'packages/...' 'packages/...'`

Comment (0xdavinchee) [reply packages/ethereum-contracts/.solcover.js:13]: didn't know that I could go in and edit it myself, but seems like I can, went ahead and made the changes  :p

Comment (shreyaspapi) [reply packages/ethereum-contracts/.solcover.js:13]: Ohh thanks 👍

**Pull request merged**